react-native/react/login2/register/set-public-name/index.js

```diff
@@ -0,0 +1,50 @@
+'use strict'
+/* @flow */
+
+import React, {Component} from '../../../base-react'
+import {setDeviceName} from '../../../actions/login2'
+import Render from './index.render'
+
+export default class SetPublicName extends Component {
+  nameTaken (deviceName) {
+    return this.props.existingDevices && this.props.existingDevices.indexOf(deviceName) !== -1
+  }
+
+  submitEnabled (deviceName) {
+    return deviceName && deviceName.length && !this.nameTaken(deviceName)
+  }
+
+  render () {
+    return (
+      <Render
+        onSubmit={ deviceName => this.props.dispatch(setDeviceName(deviceName)) }
+        nameTaken={ deviceName => this.nameTaken(deviceName) }
+        submitEnabled={ deviceName => this.submitEnabled(deviceName) }
+        {...this.props}
+      />
+    )
+  }
+
+  static parseRoute (store, currentPath, nextPath) {
+    return {
+      componentAtTop: {
+        title: '',
+        component: SetPublicName,
+        leftButtonTitle: '',
+        mapStateToProps: state => {
```

Comment:
This is the pattern that @MarcoPolo came up with with the meta-navigator. I think the driving idea is the connection is driven by the components that are pushing the things on so it's more dynamic than the static examples you traditionally see (like driven by the uri). If we can simplify it, i'm all up for it